Intermittent test failure in `TestSyntheticRadiograph.test_optical_density_histogram`

In #1456, I got a test failure for the Python 3.8 (Windows) check for `TestSyntheticRadiograph.test_optical_density_histogram`.  The test failure only showed up in one check, and did not show up in any of the other checks.  

I'm wondering if specifying a random seed for this test would enable us to resolve the problem, and also make the test deterministic.

Here's the test failure message:

```
==================================================================================== FAILURES =====================================================================================
_____________________________________________________________ TestSyntheticRadiograph.test_optical_density_histogram ______________________________________________________________
[gw0] win32 -- Python 3.8.10 D:\a\PlasmaPy\PlasmaPy\.tox\py38-all\Scripts\python.EXE

self = <plasmapy.diagnostics.tests.test_charged_particle_radiography.TestSyntheticRadiograph object at 0x0000023CE26F7AF0>

    def test_optical_density_histogram(self):
        """
        Test the optical density calculation is correct and stuffed
        with numpy.inf when the intensity is zero.
        """
        bins = (200, 60)
        size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
    
        intensity_results = cpr.synthetic_radiograph(
            self.sim_results, size=size, bins=bins
        )
        od_results = cpr.synthetic_radiograph(
            self.sim_results, size=size, bins=bins, optical_density=True
        )
    
        assert np.allclose(intensity_results[0], od_results[0])
        assert np.allclose(intensity_results[1], od_results[1])
    
        intensity = intensity_results[2]
        zero_mask = intensity == 0
        i0 = np.mean(intensity[~zero_mask])
        od = -np.log10(intensity / i0)
    
>       assert np.allclose(od[~zero_mask], od_results[2][~zero_mask])
E       assert False
E        +  where False = <function allclose at 0x0000023CD57F8B80>(array([ 5.88380294e-01,  6.50528201e-01,  3.66531544e-01,  8.59669742e-03,\n       -9.88513070e-02,  4.42252258e-01,  1...7155e+00,  4.22048872e-01, -1.27623050e-01,\n       -8.67867956e-02,  3.66531544e-01,  7.64471553e-01,  8.10229044e-01]), array([ 5.85862638e-01,  6.48010545e-01,  3.64013888e-01,  6.07904120e-03,\n       -1.01368963e-01,  4.39734602e-01,  1...5390e+00,  4.19531216e-01, -1.30140706e-01,\n       -8.93044518e-02,  3.64013888e-01,  7.61953897e-01,  8.07711387e-01]))
E        +    where <function allclose at 0x0000023CD57F8B80> = np.allclose

plasmapy\diagnostics\tests\test_charged_particle_radiography.py:598: AssertionError
```

--- a/changelog/1685.bugfix.rst
+++ b/changelog/1685.bugfix.rst
@@ -1,0 +1,1 @@
+Modified tests in the class `TestSyntheticRadiograph` to try to fix an intermittent failure of `test_optical_density_histogram`.

--- a/changelog/1685.bugfix.rst
+++ b/changelog/1685.bugfix.rst
@@ -1,1 +1,1 @@
-Modified tests in the class `TestSyntheticRadiograph` to try to fix an intermittent failure of `test_optical_density_histogram`.
+Modified tests in the class TestSyntheticRadiograph to try to fix an intermittent failure of test_optical_density_histogram.

--- a/changelog/1685.bugfix.rst
+++ b/changelog/1685.bugfix.rst
@@ -1,1 +1,1 @@
-Modified tests in the class TestSyntheticRadiograph to try to fix an intermittent failure of test_optical_density_histogram.
+Modified tests in the class ``TestSyntheticRadiograph`` to try to fix an intermittent failure of ``test_optical_density_histogram``.

--- a/plasmapy/diagnostics/tests/test_charged_particle_radiography.py
+++ b/plasmapy/diagnostics/tests/test_charged_particle_radiography.py
@@ -4,6 +4,7 @@ Tests for proton radiography functions
 
 import astropy.constants as const
 import astropy.units as u
+import copy
 import numpy as np
 import os
 import pytest
@@ -507,7 +508,7 @@ class TestSyntheticRadiograph:
         """
         Test warning when less than half the particles reach the detector plane.
         """
-        sim_results = self.sim_results.copy()
+        sim_results = copy.deepcopy(self.sim_results)
         sim_results["nparticles"] *= 3
         with pytest.warns(RuntimeWarning):
             cpr.synthetic_radiograph(sim_results)
@@ -579,22 +580,31 @@ class TestSyntheticRadiograph:
         bins = (200, 60)
         size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
 
-        intensity_results = cpr.synthetic_radiograph(
-            self.sim_results, size=size, bins=bins
-        )
+        sim_results = copy.deepcopy(self.sim_results)
+        intensity_results = cpr.synthetic_radiograph(sim_results, size=size, bins=bins)
+
+        sim_results = copy.deepcopy(self.sim_results)
         od_results = cpr.synthetic_radiograph(
-            self.sim_results, size=size, bins=bins, optical_density=True
+            sim_results, size=size, bins=bins, optical_density=True
         )
 
-        assert np.allclose(intensity_results[0], od_results[0])
-        assert np.allclose(intensity_results[1], od_results[1])
+        assert np.allclose(intensity_results[0], od_results[0], rtol=1e-4, atol=1e-7)
+        assert np.allclose(intensity_results[1], od_results[1], rtol=1e-4, atol=1e-7)
 
-        intensity = intensity_results[2]
-        zero_mask = intensity == 0
-        i0 = np.mean(intensity[~zero_mask])
-        od = -np.log10(intensity / i0)
+        # Manually calculate the OD and check that it agrees with the values
+        # returned from the function
 
-        assert np.allclose(od[~zero_mask], od_results[2][~zero_mask])
+        zero_mask = intensity_results[2] == 0
+        i0 = np.mean(intensity_results[2][~zero_mask])
+        with np.errstate(divide="ignore"):
+            od = -np.log10(intensity_results[2] / i0)
+
+        # Assert that the calculated od is close to that returned by the function
+        assert np.allclose(
+            od[~zero_mask], od_results[2][~zero_mask], rtol=1e-4, atol=1e-7
+        )
+
+        # Assert that all zero intensity values have gone to positive infinity
         assert np.all(np.isposinf(od_results[2][zero_mask]))
 
 

